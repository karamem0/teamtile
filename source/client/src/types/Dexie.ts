//
// Copyright (c) 2023 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

export interface ArrayEntity<T> {
  id: string,
  expired: number,
  values: T[]
}

export interface ValueEntity<T> {
  id: string,
  expired: number,
  value: T
}
