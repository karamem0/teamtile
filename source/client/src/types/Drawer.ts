//
// Copyright (c) 2021-2025 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

export type DrawerType =
  | 'channel'
  | 'member'
  | 'tag';
