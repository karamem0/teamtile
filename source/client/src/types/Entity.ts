//
// Copyright (c) 2021-2024 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

export enum MembershipType {
  private = 'private',
  standard = 'standard'
}

export enum VisibilityType {
  private = 'private',
  public = 'public'
}

export enum AccordionType {
  all = 'all',
  archived = 'archived',
  pinned = 'pinned'
}

export interface Channel {
  id?: string,
  displayName?: string,
  membershipType?: MembershipType,
  webUrl?: string
}

export interface Drive {
  id?: string,
  webUrl?: string
}

export interface Group {
  id?: string,
  email?: string,
  sensitivityLabel?: string
}

export interface Icon {
  id?: string,
  data?: string
}

export interface Member {
  id?: string,
  displayName?: string,
  email?: string,
  icon?: string,
  userId?: string
}

export interface Tab {
  id?: string,
  appId?: string,
  displayName?: string,
  webUrl?: string
}

export interface Team {
  id?: string,
  archived?: boolean,
  description?: string,
  displayName?: string,
  icon?: string,
  internalId?: string,
  visibility?: VisibilityType,
  webUrl?: string
}

export interface TeamInfo {
  id?: string,
  displayName?: string,
  tenantId?: string
}
