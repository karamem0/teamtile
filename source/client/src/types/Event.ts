//
// Copyright (c) 2021-2024 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';

export type Event = React.SyntheticEvent | React.SyntheticEvent<unknown> | Record<string, never>;

export type EventHandler<T = never> = (
  event?: Event,
  data?: T
) => void;
