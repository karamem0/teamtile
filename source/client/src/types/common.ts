//
// Copyright (c) 2021 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/master/LICENSE
//

export interface KeyValue<K, V> {
  key: K,
  value: V
}
