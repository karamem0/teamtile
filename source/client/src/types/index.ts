//
// Copyright (c) 2021 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/master/LICENSE
//

export interface Team {
  id?: string;
  internalId?: string;
  name?: string;
  description?: string;
  visibility?: string;
  url?: string;
  icon?: string;
}

export interface TeamChannels {
  count?: number;
  nextLink?: string;
  values?: TeamChannel[];
}

export interface TeamChannel {
  id?: string;
  name?: string;
  url?: string;
}

export interface TeamMembers {
  count?: number;
  nextLink?: string;
  values?: TeamMember[];
}

export interface TeamMember {
  id?: string;
  name?: string;
  email?: string;
}

export interface TeamDrive {
  id?: string;
  url?: string;
}
