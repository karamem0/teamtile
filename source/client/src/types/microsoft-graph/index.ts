import * as microsoftGraph from '@microsoft/microsoft-graph-types';

export type MicrosoftGraphResponse = MicrosoftGraphValueResponse | MicrosoftGraphArrayResponse | MicrosoftGraphErrorResponse | string;

export interface MicrosoftGraphValueResponse extends microsoftGraph.Entity
{
  '@odata.context': string;
}

export interface MicrosoftGraphArrayResponse {
  '@odata.context': string;
  '@odata.count': number;
  '@odata.nextLink': string;
  value: microsoftGraph.Entity[];
}

export interface MicrosoftGraphBatchResponse {
  id: string;
  status: number;
  body: MicrosoftGraphResponse;
  headers: {
    [key: string]: string;
  };
}

export interface MicrosoftGraphErrorResponse {
  error: microsoftGraph.GenericError;
}
