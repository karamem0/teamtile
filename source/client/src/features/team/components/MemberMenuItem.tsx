//
// Copyright (c) 2021-2024 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';

import {
  useAsyncFn,
  useDebounce,
  useError
} from 'react-use';

import { app } from '@microsoft/teams-js';

import { Member } from '../../../types/Entity';
import { Event } from '../../../types/Event';
import { Item } from '../../../types/Store';
import { search } from '../../../utils/String';
import { getMembers } from '../managers/TeamManager';

import Presenter from './MemberMenuItem.presenter';

interface MemberMenuItemProps {
  item?: Item
}

function MemberMenuItem(props: Readonly<MemberMenuItemProps>) {

  const { item } = props;

  const dispatchError = useError();
  const [ state, fetch ] = useAsyncFn((teamId: string) => getMembers(teamId));

  const [ values, setValues ] = React.useState<Member[]>();
  const [ filter, setFilter ] = React.useState<string>();

  React.useEffect(() => {
    if (!state.error) {
      return;
    }
    dispatchError(state.error);
  }, [
    dispatchError,
    state.error
  ]);

  useDebounce(() => {
    setValues(state.value?.filter((item) => search(item.displayName, filter)));
  }, 500, [
    state.value,
    filter
  ]);

  const handleClick = React.useCallback((_?: Event, data?: Member) => {
    if (!data?.email) {
      return;
    }
    app.openLink(`https://teams.microsoft.com/l/chat/0/0?users=${data.email}`);
  }, []);

  const handleFilterChange = React.useCallback((_?: Event, data?: string) => {
    setFilter(data);
  }, []);

  const handleOpenChange = React.useCallback(async (_?: Event, data?: boolean) => {
    if (data == null) {
      return;
    }
    if (!item?.id) {
      return;
    }
    await fetch(item.id);
  }, [
    fetch,
    item
  ]);

  return (
    <Presenter
      items={values}
      loading={state.loading}
      onClick={handleClick}
      onFilterChange={handleFilterChange}
      onOpenChange={handleOpenChange} />
  );

}

export default MemberMenuItem;
