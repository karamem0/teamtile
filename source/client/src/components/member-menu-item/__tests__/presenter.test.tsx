//
// Copyright (c) 2022 karamem0
//
// This software is released under the MIT License.
//
// https://github.com/karamem0/teamtile/blob/main/LICENSE
//

import React from 'react';

import { fireEvent, render, screen } from '@testing-library/react';

import { MemberWithIcon } from '../../../types/entity';
import MemberMenuItem from '../presenter';

beforeEach(() => {
  jest.clearAllMocks();
  jest.restoreAllMocks();
});

describe('MemberMenuItem', () => {

  it('create shapshot when menu item is not open', async () => {
    const json = await import('./__jsons__/presenter.test.json');
    const params = {
      members: json.default as MemberWithIcon[],
      onClick: jest.fn(),
      onOpen: jest.fn()
    };
    render(<MemberMenuItem {...params} />);
    expect(screen.queryAllByText(/^.*$/)[0]).toMatchSnapshot();
  });

  it('create shapshot when menu item is open', async () => {
    const json = await import('./__jsons__/presenter.test.json');
    const params = {
      members: json.default as MemberWithIcon[],
      onClick: jest.fn(),
      onOpen: jest.fn()
    };
    render(<MemberMenuItem {...params} />);
    fireEvent.click(screen.getByRole('button'));
    expect(screen.queryAllByText(/^.*$/)[0]).toMatchSnapshot();
  });

});
