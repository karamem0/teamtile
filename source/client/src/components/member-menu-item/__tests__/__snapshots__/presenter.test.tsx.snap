// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemberMenuItem create shapshot when menu item is not open 1`] = `
.emotion-0 {
  display: grid;
  grid-template-rows: auto;
  grid-template-columns: auto auto;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  cursor: pointer;
}

.emotion-1 {
  line-height: 1em;
}

.emotion-1 span {
  display: grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-2 {
  margin-left: 0.25rem;
  line-height: 0.75rem;
}

<body>
  <div>
    <span
      class="ui-text emotion-0"
      data-uses-unhanded-props="true"
      role="button"
    >
      <span
        class="emotion-1"
      >
        <span
          aria-hidden="true"
          class="root_dd790ee3"
        >
          <svg
            class="svg_dd790ee3"
            focusable="false"
            viewBox="0 0 2048 2048"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M1850 688q45 25 82 61t62 80 40 93 14 102h-128q0-52-20-99t-55-81-82-55-99-21q-53 0-99 20t-81 55-55 82-21 99q0 91-41 173t-115 136q65 33 117 81t90 108 57 128 20 142h-128q0-79-30-149t-83-122-122-82-149-31q-79 0-149 30t-122 83-82 122-31 149H512q0-73 20-141t57-128 89-108 118-82q-73-54-114-136t-42-173q0-52-20-99t-55-81-82-55-99-21q-53 0-99 20t-81 55-55 82-21 99H0q0-52 14-101t39-93 63-80 82-62q-33-35-51-81t-19-95q0-52 20-99t55-81 81-55 100-21q52 0 99 20t81 55 55 82 21 99q0 49-18 95t-52 81q82 45 134 124 54-80 138-126t182-46q97 0 181 46t139 126q52-79 134-124-33-35-51-81t-19-95q0-52 20-99t55-81 81-55 100-21q52 0 99 20t81 55 55 82 21 99q0 49-18 95t-52 81zM256 512q0 27 10 50t27 40 41 28 50 10q27 0 50-10t40-27 28-41 10-50q0-27-10-50t-27-40-41-28-50-10q-27 0-50 10t-40 27-28 41-10 50zm768 768q52 0 99-20t81-55 55-81 21-100q0-52-20-99t-55-81-82-55-99-21q-53 0-99 20t-81 55-55 82-21 99q0 53 20 99t55 81 81 55 100 21zm512-768q0 27 10 50t27 40 41 28 50 10q27 0 50-10t40-27 28-41 10-50q0-27-10-50t-27-40-41-28-50-10q-27 0-50 10t-40 27-28 41-10 50z"
            />
          </svg>
        </span>
      </span>
      <span
        class="ui-text emotion-2"
        data-uses-unhanded-props="true"
      >
        3
      </span>
    </span>
  </div>
</body>
`;

exports[`MemberMenuItem create shapshot when menu item is open 1`] = `
.emotion-0 {
  display: grid;
  grid-template-rows: auto;
  grid-template-columns: auto auto;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -ms-flex-pack: start;
  -webkit-justify-content: flex-start;
  justify-content: flex-start;
  cursor: pointer;
}

.emotion-1 {
  line-height: 1em;
}

.emotion-1 span {
  display: grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-2 {
  margin-left: 0.25rem;
  line-height: 0.75rem;
}

.emotion-3 {
  width: 12rem;
  max-height: 24rem;
  margin: -0.5rem;
  overflow: auto;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-flow: column;
  -webkit-flex-flow: column;
  -ms-flex-flow: column;
  flex-flow: column;
  gap: 0.25rem;
}

.emotion-5 {
  display: grid;
  grid-template-columns: auto auto;
  gap: 0.25rem;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: left;
  -ms-flex-pack: left;
  -webkit-justify-content: left;
  justify-content: left;
  margin: 0 -0.5rem;
}

<body>
  <div>
    <span
      class="ui-text emotion-0"
      data-uses-unhanded-props="true"
      role="button"
    >
      <span
        class="emotion-1"
      >
        <span
          aria-hidden="true"
          class="root_dd790ee3"
        >
          <svg
            class="svg_dd790ee3"
            focusable="false"
            viewBox="0 0 2048 2048"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M1850 688q45 25 82 61t62 80 40 93 14 102h-128q0-52-20-99t-55-81-82-55-99-21q-53 0-99 20t-81 55-55 82-21 99q0 91-41 173t-115 136q65 33 117 81t90 108 57 128 20 142h-128q0-79-30-149t-83-122-122-82-149-31q-79 0-149 30t-122 83-82 122-31 149H512q0-73 20-141t57-128 89-108 118-82q-73-54-114-136t-42-173q0-52-20-99t-55-81-82-55-99-21q-53 0-99 20t-81 55-55 82-21 99H0q0-52 14-101t39-93 63-80 82-62q-33-35-51-81t-19-95q0-52 20-99t55-81 81-55 100-21q52 0 99 20t81 55 55 82 21 99q0 49-18 95t-52 81q82 45 134 124 54-80 138-126t182-46q97 0 181 46t139 126q52-79 134-124-33-35-51-81t-19-95q0-52 20-99t55-81 81-55 100-21q52 0 99 20t81 55 55 82 21 99q0 49-18 95t-52 81zM256 512q0 27 10 50t27 40 41 28 50 10q27 0 50-10t40-27 28-41 10-50q0-27-10-50t-27-40-41-28-50-10q-27 0-50 10t-40 27-28 41-10 50zm768 768q52 0 99-20t81-55 55-81 21-100q0-52-20-99t-55-81-82-55-99-21q-53 0-99 20t-81 55-55 82-21 99q0 53 20 99t55 81 81 55 100 21zm512-768q0 27 10 50t27 40 41 28 50 10q27 0 50-10t40-27 28-41 10-50q0-27-10-50t-27-40-41-28-50-10q-27 0-50 10t-40 27-28 41-10 50z"
            />
          </svg>
        </span>
      </span>
      <span
        class="ui-text emotion-2"
        data-uses-unhanded-props="true"
      >
        3
      </span>
    </span>
  </div>
  <div
    class="ui-popup__content ui-animation"
    data-aa-class="Popup__popup"
    data-uses-unhanded-props="true"
    role="complementary"
    style="position: fixed; left: 0px; top: 0px; margin: 0px;"
  >
    <div
      class="ui-popup__content__content"
    >
      <div
        class="emotion-3"
      >
        <div
          class="emotion-4"
        >
          <div
            class="ui-box ui-input"
            data-uses-unhanded-props="true"
          >
            <div
              class="ui-box"
              data-uses-unhanded-props="true"
            >
              <input
                class="ui-box ui-input__input"
                data-aa-class="Input__input"
                data-uses-unhanded-props="true"
                id="ui-input-1"
                type="text"
                value=""
              />
              <div
                class="ui-box ui-input__icon"
                data-uses-unhanded-props="true"
              >
                <span
                  aria-hidden="true"
                  class="root_dd790ee3"
                >
                  <svg
                    class="svg_dd790ee3"
                    focusable="false"
                    viewBox="0 0 2048 2048"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M1344 0q97 0 187 25t168 71 142 110 111 143 71 168 25 187q0 97-25 187t-71 168-110 142-143 111-168 71-187 25q-125 0-239-42t-211-121l-785 784q-19 19-45 19t-45-19-19-45q0-26 19-45l784-785q-79-96-121-210t-42-240q0-97 25-187t71-168 110-142T989 96t168-71 187-25zm0 1280q119 0 224-45t183-124 123-183 46-224q0-119-45-224t-124-183-183-123-224-46q-119 0-224 45T937 297 814 480t-46 224q0 119 45 224t124 183 183 123 224 46z"
                    />
                  </svg>
                </span>
              </div>
            </div>
          </div>
          <ul
            class="ms-FocusZone ui-list"
            data-aa-class="List"
            data-focuszone-id="FocusZone2"
            data-uses-unhanded-props="true"
            role="menu"
          >
            <li
              class="ui-list__item"
              data-aa-class="ListItem"
              data-is-focusable="true"
              data-uses-unhanded-props="true"
              role="menuitem"
              tabindex="0"
            >
              <div
                class="ui-list__item__main"
              >
                <div
                  class="ui-list__item__headerWrapper"
                >
                  <div
                    class="ui-list__itemheader"
                    data-uses-unhanded-props="true"
                  >
                    <span
                      class="ui-text emotion-5"
                      data-uses-unhanded-props="true"
                      role="button"
                      tabindex="-1"
                    >
                      <div
                        class="ui-avatar"
                        data-uses-unhanded-props="true"
                      >
                        <span
                          class="ui-label"
                          data-uses-unhanded-props="true"
                          title="Adele Vance"
                        >
                          AV
                        </span>
                      </div>
                      <span
                        class="ui-text"
                        data-uses-unhanded-props="true"
                        dir="auto"
                      >
                        Adele Vance
                      </span>
                    </span>
                  </div>
                </div>
              </div>
            </li>
            <li
              class="ui-list__item"
              data-aa-class="ListItem"
              data-is-focusable="true"
              data-uses-unhanded-props="true"
              role="menuitem"
              tabindex="-1"
            >
              <div
                class="ui-list__item__main"
              >
                <div
                  class="ui-list__item__headerWrapper"
                >
                  <div
                    class="ui-list__itemheader"
                    data-uses-unhanded-props="true"
                  >
                    <span
                      class="ui-text emotion-5"
                      data-uses-unhanded-props="true"
                      role="button"
                      tabindex="-1"
                    >
                      <div
                        class="ui-avatar"
                        data-uses-unhanded-props="true"
                      >
                        <span
                          class="ui-label"
                          data-uses-unhanded-props="true"
                          title="Johanna Lorenz"
                        >
                          JL
                        </span>
                      </div>
                      <span
                        class="ui-text"
                        data-uses-unhanded-props="true"
                        dir="auto"
                      >
                        Johanna Lorenz
                      </span>
                    </span>
                  </div>
                </div>
              </div>
            </li>
            <li
              class="ui-list__item"
              data-aa-class="ListItem"
              data-is-focusable="true"
              data-uses-unhanded-props="true"
              role="menuitem"
              tabindex="-1"
            >
              <div
                class="ui-list__item__main"
              >
                <div
                  class="ui-list__item__headerWrapper"
                >
                  <div
                    class="ui-list__itemheader"
                    data-uses-unhanded-props="true"
                  >
                    <span
                      class="ui-text emotion-5"
                      data-uses-unhanded-props="true"
                      role="button"
                      tabindex="-1"
                    >
                      <div
                        class="ui-avatar"
                        data-uses-unhanded-props="true"
                      >
                        <span
                          class="ui-label"
                          data-uses-unhanded-props="true"
                          title="Lee Gu"
                        >
                          LG
                        </span>
                      </div>
                      <span
                        class="ui-text"
                        data-uses-unhanded-props="true"
                        dir="auto"
                      >
                        Lee Gu
                      </span>
                    </span>
                  </div>
                </div>
              </div>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</body>
`;
